analyzers/src/SonarAnalyzer.CSharp/Rules/SpecifyRouteAttribute.cs

```diff
@@ -0,0 +1,96 @@
+    protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(compilationStart =>
+            {
+                if (compilationStart.Compilation.GetTypeByMetadataName(KnownType.Microsoft_AspNetCore_Mvc_Routing_HttpMethodAttribute) is null)
+                {
+                    return;
+                }
+                compilationStart.RegisterSymbolStartAction(symbolStart =>
+                {
+                    if (symbolStart.Symbol.GetAttributes().Any(x => x.AttributeClass.Is(KnownType.Microsoft_AspNetCore_Mvc_RouteAttribute)))
+                    {
+                        return;
+                    }
+                    var secondaryLocations = new ConcurrentStack<Location>();
+                    symbolStart.RegisterSyntaxNodeAction(nodeContext =>
+                    {
+                        var node = (MethodDeclarationSyntax)nodeContext.Node;
+                        if (nodeContext.SemanticModel.GetDeclaredSymbol(node, nodeContext.Cancel) is { } method
+                            && method.IsControllerMethod()
+                            && method.GetAttributes().Any(x =>
+                                x.AttributeClass.DerivesFromAny(RouteTemplateAttributes)
+                                && x.TryGetAttributeValue<string>("template", out var template)
+                                && !CanBeIgnored(template)))
+                        {
+                            secondaryLocations.Push(node.Identifier.GetLocation());
+                        }
+                    }, SyntaxKind.MethodDeclaration);
+                    symbolStart.RegisterSymbolEndAction(symbolEnd => ReportIssues(symbolEnd, symbolStart.Symbol, secondaryLocations));
+                }, SymbolKind.NamedType);
+            });
+
+    private void ReportIssues(SonarSymbolReportingContext context, ISymbol symbol, ConcurrentStack<Location> secondaryLocations)
+    {
+        if (secondaryLocations.IsEmpty)
+        {
+            return;
+        }
+
+        foreach (var declaration in symbol.DeclaringSyntaxReferences.Select(x => x.GetSyntax()))
+        {
+            if (declaration.GetIdentifier() is { } identifier)
+            {
+                context.ReportIssue(CSharpGeneratedCodeRecognizer.Instance, Diagnostic.Create(Rule, identifier.GetLocation(), secondaryLocations));
+
+                // When a symbol was declared in multiple locations, we want to avoid reporting the same secondary locations multiple times.
+                // The list is cleared only if the declaration is not in generated code. Otherwise, the secondary locations are not reported at all.
+                if (!Language.GeneratedCodeRecognizer.IsGenerated(declaration.SyntaxTree))
+                {
+                    secondaryLocations.Clear();
+                }
```

Comment:
This is not stable between runs and may cause problems:
1. First run: secondaries are reported for primary location 1
2. Second run: secondaries are reported for primary location 2

Is this a problem on SonarCloud/SonarQube? Do they treat changes in the secondary location as a new issue?

Reply:
I do not know. The issues themselves remain the same.

Reply:
I consider this a corner case and I would not invest more time into it.

Reply:
We decided to report secondary locations everywhere. They are well handled by SQ.